Userland/Libraries/LibMarkdown/Text.h

```diff
@@ -1,54 +1,129 @@
 /*
  * Copyright (c) 2019-2020, Sergey Bugaev <bugaevc@serenityos.org>
+ * Copyright (c) 2021, Peter Elliott <pelliott@serenityos.org>
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
 #pragma once
 
 #include <AK/Noncopyable.h>
+#include <AK/NonnullOwnPtrVector.h>
+#include <AK/OwnPtr.h>
 #include <AK/String.h>
-#include <AK/Vector.h>
 
 namespace Markdown {
 
 class Text final {
-    AK_MAKE_NONCOPYABLE(Text);
-
 public:
-    struct Style {
-        bool emph { false };
-        bool strong { false };
-        bool code { false };
-        String href;
-        String img;
+    struct Node {
+        virtual void render_to_html(StringBuilder& builder) const = 0;
+        virtual void render_for_terminal(StringBuilder& builder) const = 0;
+        virtual size_t length() const = 0;
+
+        virtual ~Node() { }
+    };
+
+    struct Emph : Node {
+        bool strong;
+        NonnullOwnPtr<Node> child;
+        Emph(bool strong, NonnullOwnPtr<Node> child)
+            : strong(strong)
+            , child(move(child))
+        {
+        }
+
+        virtual void render_to_html(StringBuilder& builder) const override;
+        virtual void render_for_terminal(StringBuilder& builder) const override;
+        virtual size_t length() const override;
     };
 
-    struct Span {
+    struct Code : Node {
+        NonnullOwnPtr<Node> code;
```

Comment:
In practice code nodes can only have a multi-node of string nodes. However by keeping them as string nodes, we don't need to waste time building the whole code block into a string, and making the code more complex. I originally used a `String code'` but intentionally changed it.

Reply:
"Waste time"? I don't get it. Just read all the text there is, put it into a string builder, and done.

Reply:
Currently `parse_code` accumulates all the tokens before it knows if the code span can be closed. It seems kind of wasteful to build a string in the process, when it may fail, and doesn't really make the code any more readable IMO